rust/datafusion/src/execution/physical_plan/expressions.rs

```diff
@@ -199,6 +199,121 @@ pub fn sum(expr: Arc<dyn PhysicalExpr>) -> Arc<dyn AggregateExpr> {
+            other => Err(ExecutionError::General(format!(
+                "AVG does not support {:?}",
+                other
+            ))),
+        }
+    }
+
+    fn create_accumulator(&self) -> Rc<RefCell<dyn Accumulator>> {
+        Rc::new(RefCell::new(AvgAccumulator {
+            expr: self.expr.clone(),
+            sum: None,
+            count: None,
+        }))
+    }
+
+    fn create_combiner(&self, column_index: usize) -> Arc<dyn AggregateExpr> {
+        Arc::new(Avg::new(Arc::new(Column::new(column_index))))
+    }
+}
+
+macro_rules! avg_accumulate {
+    ($SELF:ident, $ARRAY:ident, $ROW_INDEX:expr, $ARRAY_TYPE:ident) => {{
+        if let Some(array) = $ARRAY.as_any().downcast_ref::<$ARRAY_TYPE>() {
+            if $ARRAY.is_valid($ROW_INDEX) {
+                let value = array.value($ROW_INDEX);
+                match $SELF.sum {
+                    Some(n) => {
+                        $SELF.sum = Some(n + value as f64);
+                        $SELF.count = Some($SELF.count.unwrap() + 1);
+                    }
+                    None => {
+                        $SELF.sum = Some(value as f64);
+                        $SELF.count = Some(1);
+                    }
+                };
+            }
+            Ok(())
+        } else {
+            Err(ExecutionError::General(
+                "Failed to downcast array".to_string(),
+            ))
+        }
+    }};
+}
+struct AvgAccumulator {
+    expr: Arc<dyn PhysicalExpr>,
+    sum: Option<f64>,
+    count: Option<i64>,
+}
+
+impl Accumulator for AvgAccumulator {
+    fn accumulate(&mut self, batch: &RecordBatch, row_index: usize) -> Result<()> {
+        let array = self.expr.evaluate(batch)?;
+        match self.expr.data_type(batch.schema())? {
+            DataType::Int8 => avg_accumulate!(self, array, row_index, Int8Array),
+            DataType::Int16 => avg_accumulate!(self, array, row_index, Int16Array),
+            DataType::Int32 => avg_accumulate!(self, array, row_index, Int32Array),
+            DataType::Int64 => avg_accumulate!(self, array, row_index, Int64Array),
+            DataType::UInt8 => avg_accumulate!(self, array, row_index, UInt8Array),
+            DataType::UInt16 => avg_accumulate!(self, array, row_index, UInt16Array),
+            DataType::UInt32 => avg_accumulate!(self, array, row_index, UInt32Array),
+            DataType::UInt64 => avg_accumulate!(self, array, row_index, UInt64Array),
+            DataType::Float32 => avg_accumulate!(self, array, row_index, Float32Array),
+            DataType::Float64 => avg_accumulate!(self, array, row_index, Float64Array),
+            other => Err(ExecutionError::General(format!(
+                "SUM does not support {:?}",
+                other
+            ))),
+        }
+    }
+
+    fn get_value(&self) -> Result<Option<ScalarValue>> {
+        match self.sum {
+            Some(_) => Ok(Some(ScalarValue::Float64(
+                self.sum.unwrap() / self.count.unwrap() as f64,
```

Comment:
can we avoid `unwrap`?

Reply:
The change removes the assert that if `sum` has value then `count` has too. While it's correct assumption it's kinda weird that instead of throwing it resets the accumulator. (I don't think your change shouldn't land or would worth improving, so ✔️)